crates/sui-indexer-alt-framework/src/store.rs

```diff
@@ -0,0 +1,173 @@
+use std::time::Duration;
+
+pub use scoped_futures;
+
+/// Represents a database connection that can manage watermarks for pipeline operations in the
+/// framework.
+///
+/// This trait provides methods for the necessary reads and updates for the committer, reader, and
+/// pruner components of a data pipeline, allowing for tracking progress and coordination between
+/// different pipeline stages.
+#[async_trait]
+pub trait DbConnection: Send + Sync {
+    /// Given a pipeline, return the committer watermark from the database. This is used on indexer
+    /// startup to determine which checkpoint to start or continue processing from.
+    async fn committer_watermark(
+        &mut self,
+        pipeline: &'static str,
+    ) -> anyhow::Result<Option<CommitterWatermark>>;
+
+    /// Given a pipeline, return the reader watermark from the database. This is used by the indexer
+    /// to report progress to the database.
+    async fn reader_watermark(
+        &mut self,
+        pipeline: &'static str,
+    ) -> anyhow::Result<Option<ReaderWatermark>>;
+
+    /// Upsert the high watermark as long as it raises the watermark stored in the database.
+    /// Returns a boolean indicating whether the watermark was actually updated or not.
+    async fn set_committer_watermark(
+        &mut self,
+        pipeline: &'static str,
+        watermark: CommitterWatermark,
+    ) -> anyhow::Result<bool>;
+
+    /// Update the reader low watermark for an existing watermark row, as long as this raises the
+    /// watermark, and updates the timestamp this update happened to the database's current time.
+    ///
+    /// Returns a boolean indicating whether the watermark was actually updated or not.
+    async fn set_reader_watermark(
+        &mut self,
+        pipeline: &'static str,
+        reader_lo: i64,
+    ) -> anyhow::Result<bool>;
+
+    /// Get the bounds for the region that the pruner still has to prune for the given `pipeline`,
+    /// along with a duration to wait before acting on this information, based on the time at which
+    /// the pruner last updated the bounds, and the configured `delay`. More specifically, this is
+    /// the result of delay + (pruner_timestamp - current_database_time)
+    ///
+    /// The pruner is allowed to prune the region between the returned `pruner_hi` (inclusive) and
+    /// `reader_lo` (exclusive) after `wait_for` milliseconds have passed since this response was
+    /// returned.
+    async fn pruner_watermark(
+        &mut self,
+        pipeline: &'static str,
+        delay: Duration,
+    ) -> anyhow::Result<Option<PrunerWatermark>>; // (pruner_hi, reader_lo, wait_for_ms)
+
+    /// Update the pruner watermark, returns true if the watermark was actually updated
+    async fn set_pruner_watermark(
+        &mut self,
+        pipeline: &'static str,
+        pruner_hi: i64,
+    ) -> anyhow::Result<bool>;
+}
+
+/// A storage-agnostic interface for managing pipeline watermarks.
```

Comment:
🤔

Reply:
The "managing pipeline watermarks" bit is what the Indexer cares about, but types that implement this trait will also be passed to handlers to perform arbitrary writes to the store, which is worth calling out.